From 28df8e6a7f84e4d53f6de1f9e1c977ce983368ea Mon Sep 17 00:00:00 2001
From: Normen Bolling <normen.bolling@teufel.de>
Date: Fri, 28 Oct 2016 16:43:15 +0200
Subject: [PATCH] use hw_delay_ms of asound.conf to be able to increase the the
 alsa delay

This is needed for soundbar and sounddeck.

RAD-2894
---
 pcm_hw.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/pcm/pcm_hw.c b/src/pcm/pcm_hw.c
index c34b766..6be985c 100644
--- a/src/pcm/pcm_hw.c
+++ b/src/pcm/pcm_hw.c
@@ -103,6 +103,7 @@ typedef struct {
 	snd_pcm_format_t format;
 	int rate;
 	int channels;
+	int delay_ms;
 	/* for chmap */
 	unsigned int chmap_caps;
 	snd_pcm_chmap_query_t **chmap_override;
@@ -515,6 +516,9 @@ static int snd_pcm_hw_status(snd_pcm_t *pcm, snd_pcm_status_t * status)
 		SYSMSG("SNDRV_PCM_IOCTL_STATUS failed (%i)", err);
 		return err;
 	}
+
+	status->delay = status->delay + pcm->rate * hw->delay_ms / 1000;
+
 	if (SNDRV_PROTOCOL_VERSION(2, 0, 5) > hw->version) {
 		status->tstamp.tv_nsec *= 1000L;
 		status->trigger_tstamp.tv_nsec *= 1000L;
@@ -1660,7 +1664,7 @@ int _snd_pcm_hw_open(snd_pcm_t **pcmp, const char *name,
 	long card = -1, device = 0, subdevice = -1;
 	const char *str;
 	int err, sync_ptr_ioctl = 0;
-	int rate = 0, channels = 0;
+	int rate = 0, channels = 0, hw_delay_ms = 0;
 	snd_pcm_format_t format = SND_PCM_FORMAT_UNKNOWN;
 	snd_config_t *n;
 	int nonblock = 1; /* non-block per default */
@@ -1764,6 +1768,18 @@ int _snd_pcm_hw_open(snd_pcm_t **pcmp, const char *name,
 			}
 			continue;
 		}
+		if (strcmp(id, "hw_delay_ms") == 0)
+		{
+			long val;
+			err = snd_config_get_integer(n, &val);
+			if (err < 0)
+			{
+				SNDERR("Invalid value for %s", id);
+				return err;
+			}
+			hw_delay_ms = val;
+			continue;
+		}
 		SNDERR("Unknown field %s", id);
 		snd_pcm_free_chmaps(chmap);
 		return -EINVAL;
@@ -1797,6 +1813,8 @@ int _snd_pcm_hw_open(snd_pcm_t **pcmp, const char *name,
 		hw->format = format;
 	if (channels > 0)
 		hw->channels = channels;
+	if (hw_delay_ms > 0)
+		hw->delay_ms = hw_delay_ms;
 	if (rate > 0)
 		hw->rate = rate;
 	if (chmap)
-- 
2.7.4

